### PR TITLE
MAGECLOUD-4890:  Implement support for backup databases from split connections for `ece-tools`

### PR DESCRIPTION
**Pull request by arhiopterecs**
https://magento2.atlassian.net/browse/MAGECLOUD-4890
### Description
Adds opportunity for CLI command  `db-dump`  to create backups from all available database connections

### Manual testing scenarios
https://jira.corp.magento.com/browse/MAGECLOUD-13
https://jira.corp.magento.com/browse/MAGECLOUD-151

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: src/Command/DbDump.php [hunk 1]
```diff
@@ -135,8 +147,10 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $this->maintenanceModeSwitcher->enable();
             $this->cronSwitcher->disable();
             $this->backgroundProcess->kill();
-            $this->dumpGenerator->create((bool)$input->getOption(self::OPTION_REMOVE_DEFINERS));
-            $this->logger->info('Backup completed.');
```
Comment (oshmyheliuk) [inline src/Command/DbDump.php:139]: Why did you remove info log about backup completion?

File: src/Config/Validator/Deploy/ResourceConfiguration.php [hunk 2]
```diff
@@ -45,17 +54,26 @@ public function __construct(
...
     public function validate(): Validator\ResultInterface
     {
+        $dbConfig = $this->dbConfig->get();
+        $resourceConfig = $this->resourceConfig->get();
         $wrongResources = [];
-        foreach ($this->resourceConfig->get() as $resourceName => $resourceData) {
-            if (!isset($resourceData['connection'])) {
+        foreach ($resourceConfig as $resourceName => $resourceData) {
+            if (!isset($resourceData[ResourceConfig::KEY_CONNECTION])
+                || !isset($dbConfig[DbConfig::KEY_CONNECTION][$resourceData[ResourceConfig::KEY_CONNECTION]])) {
```
Comment (oshmyheliuk) [inline src/Config/Validator/Deploy/ResourceConfiguration.php]: Please close multiline statements on the new line. For better code readiness.

File: src/DB/Data/ConnectionFactory.php [hunk 3]
```diff
@@ -41,21 +43,41 @@ public function create(string $connectionType): ConnectionInterface
...
+                return $this->getConnectionData(DbConfig::CONNECTION_SALES, false);
             default:
                 throw new \RuntimeException(
                     sprintf('Connection with type %s doesn\'t exist', $connectionType)
                 );
         }
+    }
 
-        return $connection;
+    private function getConnectionData($name, $isMain = true): ConnectionInterface
```
Comment (oshmyheliuk) [inline src/DB/Data/ConnectionFactory.php]: Please add method description

File: src/DB/Data/ConnectionFactoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MagentoCloud\DB\Data;
+
+/**
+ * Interface of database connection data factory
+ */
+interface ConnectionFactoryInterface
```
Comment (oshmyheliuk) [inline src/DB/Data/ConnectionFactoryInterface.php]: What the purpose of this interface if you don't use it anywhere?

File: src/DB/DumpGenerator.php [hunk 5]
```diff
@@ -7,19 +7,52 @@
...
  */
 class DumpGenerator
 {
     /**
-     * Template for dump file name where %s should be changed to timestamp for uniqueness
+     * Database names
      */
-    const DUMP_FILE_NAME_TEMPLATE = 'dump-%s.sql.gz';
+    const DATABASE_MAIN = 'main';
+    const DATABASE_QUOTE = 'quote';
```
Comment (oshmyheliuk) [inline src/DB/DumpGenerator.php]: You already added 'quote' and 'sales' constants in DeployInterface is better not duplicate constants with the same value.  `const DATABASE_QUOTE = DeployInterface::SPLIT_DB_VALUE_QUOTE;`

File: src/DB/DumpGenerator.php [hunk 6]
```diff
@@ -106,35 +237,37 @@ public function create(bool $removeDefiners)
...
-                }
-                flock($lockFileHandle, LOCK_UN);
-            } else {
+            if (!flock($lockFileHandle, LOCK_EX)) {
                 $this->logger->info('Dump process is locked!');
+                return;
             }
-            fclose($lockFileHandle);
-        } catch (\Exception $e) {
-            $this->logger->error($e->getMessage());
```
Comment (oshmyheliuk) [inline src/DB/DumpGenerator.php:137]: Why did you remove error logging?

File: src/DB/DumpGenerator.php [hunk 7]
```diff
@@ -54,39 +87,137 @@ class DumpGenerator
...
      * as well as serves a log with the name of created dump file.
      * If any error happened during dumping, dump file is removed.
      *
+     * @param string $database
+     * @param ConnectionInterface $connectionData
      * @param bool $removeDefiners
-     * @return void
-     * @throws \Magento\MagentoCloud\Package\UndefinedPackageException
+     * @throws UndefinedPackageException
      */
```
Comment (oshmyheliuk) [inline src/DB/DumpGenerator.php:99]: ShellException can be thrown as well after you removed the catch statement.

File: src/DB/DumpGenerator.php [hunk 8]
```diff
@@ -106,35 +237,37 @@ public function create(bool $removeDefiners)
...
+
+            $command = 'timeout ' . self::DUMP_TIMEOUT . ' ' . $this->dump->getCommand($connectionData);
+            if ($removeDefiners) {
+                $command .= ' | sed -e \'s/DEFINER[ ]*=[ ]*[^*]*\*/\*/\'';
+            }
+            $command .= ' | gzip > ' . $dumpFile;
+
+            $process = $this->shell->execute('bash -c "set -o pipefail; ' . $command . '"');
+
+            if ($process->getExitCode() !== ShellInterface::CODE_SUCCESS) {
```
Comment (bbatsche) [inline src/DB/DumpGenerator.php]: `execute()` uses Symfony's `mustRun()` method, so there's no way `getExitCode()` could return anything other than success without throwing an exception

File: src/DB/DumpGenerator.php [hunk 9]
```diff
@@ -106,35 +237,37 @@ public function create(bool $removeDefiners)
...
+                $this->logger->error('Error has occurred during mysqldump');
+                $this->shell->execute('rm ' . $dumpFile);
```
Comment (bbatsche) [inline src/DB/DumpGenerator.php]: `$dumpFile` may not exist if the command has failed so removing it will cause a second failure

**Review by bbatsche — CHANGES_REQUESTED** [review]

File: src/DB/DumpGenerator.php (hunk 8 above)
Comment (arhiopterecs) [reply src/DB/DumpGenerator.php]: Fixed

File: src/DB/DumpGenerator.php (hunk 9 above)
Comment (arhiopterecs) [reply src/DB/DumpGenerator.php]: Fixed

File: src/DB/DumpGenerator.php (hunk 7 above)
Comment (arhiopterecs) [reply src/DB/DumpGenerator.php:99]: Added

File: src/DB/DumpGenerator.php (hunk 6 above)
Comment (arhiopterecs) [reply src/DB/DumpGenerator.php:137]: @oshmyheliuk  error logging will occur in [Magento\MagentoCloud\Command\ DbDump::execute()](https://github.com/magento/ece-tools/pull/673/files#diff-c1669aa33365b724fea44b5e9261d06cR155)

File: src/DB/DumpGenerator.php (hunk 5 above)
Comment (arhiopterecs) [reply src/DB/DumpGenerator.php]: @oshmyheliuk These constants have different contexts

File: src/DB/Data/ConnectionFactoryInterface.php (hunk 4 above)
Comment (arhiopterecs) [reply src/DB/Data/ConnectionFactoryInterface.php]: I have removed the interface for now.

File: src/DB/Data/ConnectionFactory.php (hunk 3 above)
Comment (arhiopterecs) [reply src/DB/Data/ConnectionFactory.php]: Done

File: src/Config/Validator/Deploy/ResourceConfiguration.php (hunk 2 above)
Comment (arhiopterecs) [reply src/Config/Validator/Deploy/ResourceConfiguration.php]: Done

File: src/Command/DbDump.php (hunk 1 above)
Comment (arhiopterecs) [reply src/Command/DbDump.php:139]: It seems like an error occurred while resolving conflicts. I returned the string back

**Review by bbatsche — DISMISSED** [review]

**Review by bbatsche — APPROVED** [review]

File: src/Command/DbDump.php [hunk 10]
```diff
@@ -132,19 +138,58 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+     * @throws ConfigException
+     */
+    private function checkConnectionsAvailability(array $connections): bool
+    {
+        $result = true;
+        $envConnections = $this->dbConfig->get()[DbConfig::KEY_CONNECTION] ?? [];
+        foreach ($connections as $connection) {
+            if (!isset($envConnections[DumpProcessor::CONNECTION_MAP[$connection]])) {
+                $this->logger->error(sprintf(
+                    'Environment has not connection `%s` associated with database `%s`',
```
Comment (NadiyaS) [inline src/Command/DbDump.php:186]: `has not` -> does not have

File: src/Command/DbDump.php [hunk 11]
```diff
@@ -91,8 +70,17 @@ public function __construct(
...
-            ->setDescription('Creates backup of database');
-
+            ->setDescription('Creates backups of databases');
+        $this->addArgument(
+            self::ARGUMENT_DATABASES,
+            InputArgument::IS_ARRAY,
+            sprintf(
+                'Databases to backup. Available values: %s or empty. By default will backup the databases'
+                . ' based on the databases configuration from the file <magento_root>/app/etc/env.php ',
+                implode(' ', DumpProcessor::DATABASE_MAP)
```
Comment (NadiyaS) [inline src/Command/DbDump.php:80]: Why not to leave IDs from env.php: default, checkout, sales?

File: src/DB/DumpProcessor.php [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        ConnectionFactory::CONNECTION_SLAVE => self::DATABASE_MAIN,
+        ConnectionFactory::CONNECTION_QUOTE_SLAVE => self::DATABASE_QUOTE,
+        ConnectionFactory::CONNECTION_SALES_SLAVE => self::DATABASE_SALES,
+    ];
+
+    /**
+     * Connection map:
+     * {connection name from environment} => {magento connection name}
+     */
+    const CONNECTION_MAP = [
```
Comment (NadiyaS) [inline src/DB/DumpProcessor.php:46]: there are too many maps related to split DB. I propose to create one class with one map where all other classes will retrieve ids for env.php or for Split_DB

File: src/Command/DbDump.php [hunk 13]
```diff
@@ -91,8 +70,17 @@ public function __construct(
...
     {
         $this->setName(self::NAME)
-            ->setDescription('Creates backup of database');
-
+            ->setDescription('Creates backups of databases');
+        $this->addArgument(
+            self::ARGUMENT_DATABASES,
+            InputArgument::IS_ARRAY,
+            sprintf(
+                'Databases to backup. Available values: %s or empty. By default will backup the databases'
```
Comment (NadiyaS) [inline src/Command/DbDump.php:78]: values should be framed with quotes

File: src/Command/DbDump.php (hunk 11 above)
Comment (NadiyaS) [reply src/Command/DbDump.php:80]: also, implode values through comma would be more readable

File: src/Command/DbDump.php (hunk 10 above)
Comment (arhiopterecs) [reply src/Command/DbDump.php:186]: Done

File: src/Command/DbDump.php (hunk 11 above)
Comment (arhiopterecs) [reply src/Command/DbDump.php:80]: > Why not to leave IDs from env.php: default, checkout, sales?

Because console commands of Magento use the part like 'sales'  and 'quote'​​to switch to split database mode:
```
magento setup:db-schema:split-sales --host="<checkout db host or ip>" --dbname="<name>" --username="<checkout db username>" --password="<password>"
magento setup:db-schema:split-quote --host="<checkout db host or ip>" --dbname="<name>" --username="<checkout db username>" --password="<password>"

```
Also, the same values ​​are used to enable split database mode on Magento Cloud:
```
stage:
  deploy:
    SPLIT_DB:
       - quote
       - sales
```

> also, implode values through comma would be more readable

Yes, but it can be confusing for the user because the values ​​are entered using a space as a separator

File: src/DB/DumpProcessor.php (hunk 12 above)
Comment (arhiopterecs) [reply src/DB/DumpProcessor.php:46]: If to create one map, it will be a nested array. Anyway, working with single-level arrays is easier.
To access the values ​​of such a structure, I will most likely need nested constructions "foreach", which is not entirely good and|or additional methods for access to those values.
And there are only two maps

File: src/Command/DbDump.php (hunk 13 above)
Comment (arhiopterecs) [reply src/Command/DbDump.php:78]: Done

**Comment by arhiopterecs** [comment]
This PR was recreated https://github.com/magento/ece-tools/pull/687

File: src/DB/DumpProcessor.php (hunk 12 above)
Comment (NadiyaS) [reply src/DB/DumpProcessor.php:46]: I did not mean having one array, - just one place to resolve such mappings. Right now we have 4 mapping arrays related to DB in 3 classes plus these 2 you will add.
